File: k_diff_pairs_array.cpp
//Time Complexity - O(nlogn)
//Space complexity - O(1)
//Ran successfully on leetcode
/*3 point algo is as follows - 
	1. We sort the array first and then start two pointers slow and fast from the beginning
	2. Then we compare the difference of slow and fast pointers. If its greater than k, then we move the slow pointer. Else, we move the fast pointer.
	3. If the difference is equal to k, then we increase count and move both pointers. 
	4. Then we interate the fast pointer if the previous element is the same, in order to avoid duplicates.*/

class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        if(nums.empty())
            return 0;
        sort(nums.begin(),nums.end());
        int slow=0,fast=1,count=0;
        while(fast<nums.size())
        {
            if(slow==fast || nums[fast]-nums[slow]<k)
                fast++;
            else if(nums[fast]-nums[slow]>k)
                slow++;
            else
            {
                count++;
                slow++;
                fast++;
                while(fast<nums.size() && nums[fast]==nums[fast-1])
                    fast++;
            }
        }
        return count;
    }
};

File: k_diff_pairs_array.java
//TimeComplexity: O(n)

class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i =0; i<nums.length;i++) {
            if(map.containsKey(nums[i])) {
               map.put(nums[i], map.get(nums[i])+1);
            } else {
                map.put(nums[i],1);
            }
        }
       int findcount=0;
        for(int i : map.keySet()) {
            if(map.containsKey(i-k) && k>0) {
                findcount++;
            } else if(map.get(i)>1 && k ==0) {
                findcount++;
            }
          
            }
            return findcount;
        }
}
        
    
